### PR TITLE
[CH-45]Support count(1)/count(*)

### PR DESCRIPTION
**Pull request by lgbo-ustc**
Relate to the issue https://github.com/Kyligence/ClickHouse/issues/45

## What changes were proposed in this pull request?

A empty base_schema will be made in the `scan` node  when run the following queries
```SQL
select count(1) from t;
select count(*) from t;
select sum(2) from t;
select 1 from t;
``` 

The backend source node will return empty data or throw an exception when there is no column to read.
So we fill a attribute to read in `FileSourceScanExec`. 

Another problem is in `ProjectExecTransformer`. If `ProjectExecTransformer`'s child is an `ColumnarInputAdapter` and `projectList` is empty, it would generate a scan node with empty base_schema. So when `childCtx` is null, we don't return directly.

## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)

(If this patch involves UI changes, please attach a screenshot; otherwise, remove this)

### CODE REVIEW

File: jvm/src/main/scala/io/glutenproject/extension/ColumnarOverrides.scala [hunk 1]
```diff
@@ -70,9 +72,18 @@ case class TransformPreOverrides() extends Rule[SparkPlan] {
         new BatchScanExecTransformer(plan.output, plan.scan, newPartitionFilters)
       case plan: FileSourceScanExec =>
         logDebug(s"Columnar Processing for ${plan.getClass} is currently supported.")
+        val needToFillAttr = conf.getConfString(GlutenConfig.GLUTEN_FILL_ATTRIBUTE_FOR_SCAN,
+            "true").toBoolean
+        val realOutput = if (plan.output.size == 0 && needToFillAttr) {
```
Comment (zzcclp) [inline jvm/src/main/scala/io/glutenproject/extension/ColumnarOverrides.scala:77]: It's better to use SPI API to support this future for ClickHouse Backend, Velox Backend does not need this.

**Comment by zzcclp** [comment]
Please add some UTs

File: jvm/src/main/scala/io/glutenproject/extension/ColumnarOverrides.scala [hunk 2]
```diff
@@ -66,9 +68,11 @@ case class TransformPreOverrides() extends Rule[SparkPlan] {
         new BatchScanExecTransformer(plan.output, plan.scan, newPartitionFilters)
       case plan: FileSourceScanExec =>
         logDebug(s"Columnar Processing for ${plan.getClass} is currently supported.")
+        val realOutput = BackendsApiManager.getTransformerApiInstance.
```
Comment (zzcclp) [inline jvm/src/main/scala/io/glutenproject/extension/ColumnarOverrides.scala:71]: what about BatchScanExecTransformer?

**Comment by zzcclp** [comment]
In `FilterHandler.applyFilterPushdownToScan` , it also needs to handle like this.

BTW, please rebase to main

File: jvm/src/main/scala/io/glutenproject/extension/ColumnarOverrides.scala (hunk 2 above)
Comment (lgbo-ustc) [reply jvm/src/main/scala/io/glutenproject/extension/ColumnarOverrides.scala:71]: How to trigger the BatchScanExecTransformer

**Comment by lgbo-ustc** [comment]
I don't think we need to add this in `FilterHandler.applyFilterPushdownToScan`. In this case, the columns to read are never empty

File: jvm/src/main/scala/io/glutenproject/backendsapi/ITransformerApi.scala [hunk 3]
```diff
@@ -44,4 +45,11 @@ trait ITransformerApi extends IBackendsApi {
    */
   def genInputPartitionSeq(relation: HadoopFsRelation,
                            selectedPartitions: Array[PartitionDirectory]): Seq[InputPartition]
+
+  /**
+   * Clickhouse doesn't support read from a file source with empty attributes now.
+   */
+  def genFileSourceScanAttributes(scanNode: FileSourceScanExec): Seq[Attribute] = {
```
[the remark this answers is not in the document]
Comment (zzcclp) [reply jvm/src/main/scala/io/glutenproject/backendsapi/ITransformerApi.scala:52]: Use the default implementation

File: jvm/src/main/scala/io/glutenproject/extension/ColumnarOverrides.scala (hunk 2 above)
Comment (zzcclp) [reply jvm/src/main/scala/io/glutenproject/extension/ColumnarOverrides.scala:71]: Please refer to the test case: GlutenClickHouseDSV2Suite

**Comment by lgbo-ustc** [comment]
wait for https://github.com/Kyligence/ClickHouse/pull/175 be merged

File: jvm/src/main/scala/io/glutenproject/execution/BasicPhysicalOperatorTransformer.scala [hunk 4]
```diff
@@ -397,8 +394,10 @@ case class ProjectExecTransformer(projectList: Seq[NamedExpression],
         null
     }
     val operatorId = context.nextOperatorId
-    if (projectList == null || projectList.isEmpty) {
+    if ((projectList == null || projectList.isEmpty) && childCtx != null) {
```
[the remark this answers is not in the document]
Comment (lgbo-ustc) [reply jvm/src/main/scala/io/glutenproject/execution/BasicPhysicalOperatorTransformer.scala:397]: This case would happend?

**Review by rui-mo — APPROVED** [review]
Verified on Jenkins. LGTM.

**Pull request merged**